llvm/tools/sycl-post-link/ModuleSplitter.cpp

```diff
@@ -291,6 +335,14 @@ ModuleDesc extractSubModule(const ModuleDesc &MD,
   // declarations and removed later.
   std::unique_ptr<Module> SubM = CloneModule(
       M, VMap, [&](const GlobalValue *GV) { return GVs.count(GV); });
+  // Original module may have named metadata (spirv.ExecutionMode) referencing
+  // kernels in the module. Some of the Metadata nodes may reference kernels
+  // which are not included into the extracted submodule, in such case
+  // CloneModule either leaves that metadata nodes as is but they will reference
+  // dead prototype of the kernel or operand will be replace with nullptr. So
+  // process all nodes in the named metadata and remove nodes which are
+  // referencing kernels which are not included into submodule.
+  processSubModuleNamedMetadata(SubM.get());
```

Comment:
Should this be moved into cleanup "phase" we have, which is performed for each module after split?

https://github.com/intel/llvm/blob/3245ce93accea25b87d5c6f90da0e099858b8312/llvm/tools/sycl-post-link/ModuleSplitter.cpp#L623-L632

Reply:
I was thinking of adding this there but there is a TODO comment to remove ModuleDesc::cleanup(), that's why I decided to keep introduced function out of it.

Reply:
If I understand correctly the intent of the comment is to move that particular _call_ to `cleanup`, because we do call that function later as well. I don't think that we intent to remove the cleanup routine all together. Tagging @asudarsa to correct me if I'm wrong

Reply:
Got it, thank you, moved to `cleanup`.